reflect/src/main/java/dagger/reflect/Key.java

```diff
@@ -20,19 +20,39 @@
 
 import com.google.auto.value.AutoValue;
 import java.lang.annotation.Annotation;
+import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
 import org.jetbrains.annotations.Nullable;
 
 @AutoValue
 abstract class Key {
   static Key of(@Nullable Annotation qualifier, Type type) {
+    if (containsTypeVariable(type)) {
```

Comment:
@JakeWharton I can remove this check if it is causing any second thoughts on merging this PR.
It helps surface unsupported bindings in Dagger Reflect as soon as possible. When a Key which includes `<T>` instead of a real type, it immediately throws an error.
It has a bit of a runtime cost since it recursively checks the type and its parameters whenever a Key is constructed.